### PR TITLE
Average ages

### PR DESCRIPTION
**Pull request by vsupruniuk**

### CODE REVIEW

**Review by jstmpelowycz — CHANGES_REQUESTED** [review]

**Comment by vsupruniuk** [comment]
I will try to fix the problem of time complexity when I finish the course of algorithms and data structures

File: src/averageAges.js [hunk 1]
```diff
@@ -14,12 +14,21 @@
...
  */
+
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with &&, || or ?:
-  // without nesting
+  const men = people.filter(person => person.sex === 'm');
+  const menInCentry = century
```
Comment (Olexandr6) [inline src/averageAges.js:20]: It's not `menInCentry` in cause `century` === `undefined`

File: src/averageAges.js [hunk 2]
```diff
@@ -14,12 +14,21 @@
...
+    ? men.filter(
+      person => Math.ceil(person.died / 100) === century
+    )
+    : men;
```
Comment (Olexandr6) [inline src/averageAges.js:24]: You filter 2 times when there is `century`. Let's implement this using only one filter

File: src/averageAges.js [hunk 3]
```diff
@@ -37,7 +46,22 @@ function calculateMenAverageAge(people, century) {
...
+  const women = people.filter(person => person.sex === 'f');
+
+  const womenWithChild = withChildren
+    ? people.filter(
+      person => people.some(
+        child => child.mother === person.name
+      )
+    )
+    : women;
+
```
Comment (Olexandr6) [inline src/averageAges.js:58]: The same case here, let's use filter only once

File: src/averageAges.js [hunk 4]
```diff
@@ -55,7 +68,17 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const child = people
+    .filter(children => people.some(person => children.mother === person.name));
```
[the remark this answers is not in the document]
Comment (Olexandr6) [reply src/averageAges.js:83]: HashMap is crying🥲

File: src/averageAges.js [hunk 5]
```diff
@@ -55,7 +79,23 @@ function calculateWomenAverageAge(people, withChildren) {
...
+
+  const childIsSon = onlyWithSon
+    ? child.filter(
+      person => person.sex === 'm'
+    )
+    : child;
```
Comment (Olexandr6) [inline src/averageAges.js:89]: Let's use filtering once🥲

File: src/averageAges.js [hunk 6]
```diff
@@ -55,7 +79,23 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const child = people
+    .filter(children => people.some(person => children.mother === person.name));
+
+  const childIsSon = onlyWithSon
```
Comment (Olexandr6) [inline src/averageAges.js:85]: It's not always sons here. Pay attention to variables naming
```suggestion
  const childIsSon = onlyWithSon
```

**Review by G-YaSha — CHANGES_REQUESTED** [review]
Almost done

**Review by G-YaSha — APPROVED** [review]

**Pull request closed without merging**